File: src/bot/embeds.ts
import { Colors, EmbedBuilder } from 'discord.js';
import { Article } from '../types/wordpress/Article';

const MAX_DESC = 2047;

function getEmbed(): EmbedBuilder {
  return new EmbedBuilder();
}

export function getPostEmbed(article: Article): EmbedBuilder {
  const getDesc = () => {
    let result = article.yoast_head_json.og_description;
    if (result.length > MAX_DESC) {
      const suffix = `... [View more here!](${article.yoast_head_json.canonical})`;
      const end = (MAX_DESC - 1) - suffix.length;
      const newDesc = result.substring(0, end);
      result = newDesc + suffix;
    }
    return result;
  };
  return getEmbed()
    .setTitle(article.yoast_head_json.og_title)
    .setDescription(getDesc())
    .setColor(Colors.Blue)
    .setURL(article.yoast_head_json.canonical)
    .setThumbnail(article.yoast_head_json.og_image[0]?.url || null)
    .setAuthor({
      name: article.yoast_head_json.author,
    })
    .setTimestamp(new Date(article.date_gmt));
}

export function getErrorEmbed(e: unknown): EmbedBuilder {
  const err = e instanceof Error ? e.message : `${e}`;
  const desc = `\`\`\`\n${err}\n\`\`\``;
  return getEmbed()
    .setColor(Colors.Red)
    .setTitle('An Error has Occurred')
    .setDescription(desc);
}

File: src/bot/embeds.rs
use crate::types::Article;
use htmlentity::entity::*;
use serenity::builder::CreateEmbed;
use std::collections::HashMap;
use substring::Substring;

const MAX_DESC: usize = 2047;

fn fix_description(url: String, desc: &String) -> String {
    let mut result = String::new();
    let length = desc.len();

    decode_to(desc, &mut result);
    if length > MAX_DESC {
        let suffix = format!("... [View more here!]({})", url);
        let end = (MAX_DESC - 1) - suffix.len();
        let new_desc = result.substring(0, end);

        result = String::from(format!("{}{}", new_desc, suffix));
    }

    result
}

impl Into<CreateEmbed> for Article {
    fn into(self) -> CreateEmbed {
        let mut builder = CreateEmbed(HashMap::new());

        if let Some(json) = self.yoast_head_json {
            let url = json.canonical;
            if let Some(title) = json.og_title {
                builder.title(title);
            }
            if let Some(description) = json.og_description {
                let new_desc = fix_description(url, &description);
                builder.description(new_desc);
            } else {
                builder.description("description");
            }
        }

        builder
    }
}
